### ISSUE TITLE
Accessing HLS object from player instance

### ISSUE THREAD

**Issue by talk2cerlin** [streamroot/videojs-hlsjs-plugin#20]
if `autoStartLoad` is set to false, according to [this](https://github.com/dailymotion/hls.js/blob/master/API.md#fine-tuning) document i will have to make an explicit call to start loading ie) `hls.startLoad(startPosition=-1)` . for this i need the underlying hls instance. 

is there any way i can get that instance?

**Comment by aquu** [comment]
I'm also wondering if this is possible? I'm trying to listen events on Hls object

**Comment by tchakabam** [comment]
hey! currently not possible. videojs wraps the underlying media engines in an opaque way.

one could say that it kind off defies the purpose of such a layer as videojs to expose the underlying things, as the idea of the media player is to abstract away from how the media is played in a specific case.

imho the correct solution would be to implement analoguous properties/options/events in videojs, and then pass them on to the underlying Hls.js instance if need-be.

Is there an autoStart option in videojs that we could pass on to the hls instance, or is there a some functionnality like `startLoad` on videojs?

@aquu which events of hls.js are you interested in? usually in videojs one propagates the events to the `tech` object (as it is done in our wrapper already for error events)

**Comment by tchakabam** [comment]
@talk2cerlin @aquu There would be a possibility however to expose the hls.js instance via the `tech` object of the videojs instance directly (I assume). So that is always an option probably to do this somehow.

But the preferred way I think would be to perform proper abstraction as described above.

**Comment by aquu** [comment]
@tchakabam Hi, thanks for response and clarification. I'm not sure if it is possible to get this working even if I could access Hls object.

What I'm trying to find out way is to listen timed metadata from live stream. When using [videojs-contrib-hls](https://github.com/videojs/videojs-contrib-hls) it was possible like this but for some reason I can't get addtrack events to fire when using videojs5-hlsjs-source-handler. Is this even possible with your source handler?

`
    player.textTracks().addEventListener('addtrack', function(addTrackEvent) {
      var track = addTrackEvent.track;
      track.mode = 'hidden';
      track.addEventListener('cuechange', function(cueChangeEvent) {
      });
    });
`

**Comment by tchakabam** [comment]
@aquu Hey, Yes indeed we should be able to implement any kind off event that Hls.js can deliver.

I think Hls.js doesn't really implement timed-metadata yet (i am not sure though). You should probably first try to find an approach based solely on Hls.js that implements your basic functionnality of dealing with text-tracks (similar to the example above but with Hls.js). It is possible that this is a missing feature there. So it would have to be added first.

And then when it's clear what feature of Hls.js we want to bring in the videojs world, I am confident it will be possible to be added to our videojs source-handler here. 

Does that make sense to you or did i miss something?

**Comment by aquu** [comment]
@tchakabam Yeah, thanks for your response. That makes sense, and yes I will try to find out first if timed metadata is possible at all with Hls.js. And I hope it is (or will be) possible because in our use case timing other content to stream is important.

**Comment by tchakabam** [comment]
@aquu yes, right. but then again i can't see any working demo off it. and looking at the code itself i am not sure if this stuff is really hooked up. there are some parsers for various text formats etc, but not sure if its kind off a "dorming feature" - will try to find out ;)

**Comment by tchakabam** [comment]
@aquu ok so forget what i just said, the feature seems to be implemented fully and expected to work. do you even see the captions displayed in a barebones player setup (or with videojs) using hls.js on your stream? can you post a test stream?

**Comment by aquu** [comment]
@tchakabam I didn't yet have time for testing this yet with hls.js. I will look at this later and maybe able to put test stream up if you want to test it. (Now I have to manually inject ID3 tags to the stream for testing.)

If you found any examples, please post them here?

**Comment by woutd** [comment]
I would like to set **startLevel** (to -1), but this is currently not possible since it is not a configuration parameter, and can only be set with the API. Any advice how to do this?

**Comment by tchakabam** [comment]
@woutd Yes that would be nice. You should probably post that as a separate issue as a feature request. 

You could also implement it yourself, which would require one change here and one in hls.js to be able to pass that as a config param. I think those would be simple changes in both projects that would probably get merged quite quickly!

Currently hls.js config doesn't contain that parameter, but one could add it (see https://github.com/dailymotion/hls.js/blob/master/src/hls.js#L53)

Then you could pass `startLevel` via the `hlsjsConfig` field like other params, see example: https://github.com/streamroot/videojs5-hlsjs-source-handler/blob/master/debug.html#L28

**Comment by tchakabam** [comment]
@aquu sorry missed your reply there back then. i m still note sure what you re really doing to be honest. if you provide a way that i can reproduce what you're doing with hls.js exactly, i might be able to help you in order to integrate it further with this project or others.

**Comment by woutd** [comment]
Thanks for the info @tchakabam 
I issued a PR to add a startLevel configuration parameter in hls.js.

**Comment by aquu** [comment]
@tchakabam I finally had time to dig in deeper and found out how to do what I want.

To listen timed-metadata, you have to listen  ```Hls.Events.FRAG_PARSING_METADATA ``` [documented here](https://github.com/dailymotion/hls.js/blob/v0.5.x/API.md#runtime-events)

Then I just added following listener to your source handler code here https://github.com/streamroot/videojs5-hlsjs-source-handler/blob/f26b81a9f3b184374952c22fa06eb0283915f14e/lib/videojs5-hlsjs-source-handler.js#L41

 ```
_hls.on(Hls.Events.FRAG_PARSING_METADATA,function(event,data) {
    var cuepoint = new TextDecoder("utf-8").decode(data.samples[0].data);
    console.log(cuepoint); 
 });
 ```
So is there a way attach listeneres for Hls.js runtime events via hlsjsConfig or some other way to listen those events?

**Comment by tchakabam** [comment]
@aquu thanks for providing this usecase. it probably needs to be integrated with videojs in an appropriate way. probably one needs to call `addTextTrack` on the HTML5-tech object, as here [1] ?

might take a close look later, but basically the idea should be to pass on either the TextDecoder or the raw data to video js components so that it can take care of displaying that, right?

[1] https://github.com/videojs/video.js/blob/master/src/js/tech/html5.js#L579

**Comment by aquu** [comment]
Actually I have other functionality outside the player that uses timed cuepoints from the live stream. But yes, calling addTextTrack could be easiest option to implement this with Hls.js. With videojs-contrib-hls cuepoints are automatically inserted to TextTracks so I can listen directly TextTrack events like I explained [earlier](https://github.com/streamroot/videojs5-hlsjs-source-handler/issues/20#issuecomment-240099130).

**Comment by tchakabam** [comment]
@aquu ok i get you now. `addtrack` event gets triggered when you call `addTrack` on a TrackList in videojs. so that happens probably when calling the above mentioned `addTextTrack`.

so once you do that, you should get the same behavior as you get with `videojs-contrib-hls`.

probably looking at the source-code of the latter should help you to just align this integration here with it (they should basically translate the Hls.js events into videojs tech calls in the same way).

Now if you want to expose the cue-points to the application directly, that would probably need another event that may or may not exist yet in videojs that would carry that data pushed for every single cue-point decoded. Is that also already possible with `videojs-contrib-hls`?

**Comment by tchakabam** [comment]
this seems to be how its being done in `contrib-hls`: https://github.com/videojs/videojs-contrib-hls/blob/77c25202ad91ad8ecf347a143d0e4276d46a4883/src/master-playlist-controller.js#L195

**Comment by aquu** [comment]
@tchakabam Yeah, that is what I meant. And yes, cueChangeEvent only tells when cue changes, but does not tell us which cuepoint just fired. I solved that problem by looping over all cues (this.cues) and comparing currentTime and cue.startTime to find yout current cue. 

I didn't yet have time to test this out with Hls.js but its looks promising :) 

Anyway, more related to original issue: There is currently no way to listen [Hls.js events](https://github.com/dailymotion/hls.js/blob/v0.5.x/API.md#runtime-events)? (Other than putting it directly to your source handler code, like I just did)

**Comment by tchakabam** [comment]
@aquu let me turn around your question? from a videojs point of view, how would you imagine being able to listen Hls.js events?

exposing Hls.js specific things somehow breaks the idea of a generic player API that video.js should be.

so as i think i already suggested, the right way to do things is translating each Hls.js event (listeneing and re-triggering) into corresponding videojs events, and then have the application attach respective listeners to the videojs player instance.

How else would you imagine this to be supposed to work?

**Comment by aquu** [comment]
Hmm, you are probably right. I was just wondering if it would be good idea to expose Hls.js events directly via player configuration or something, (But you already pointed out thing about generic player API and it makes sense)

**Comment by tchakabam** [comment]
@aquu plus configuration doesn't seem an appropriate place to pass callbacks, that's what events are for :) 

making proper videojs things and making the underlying media-engine opaque seems the right way to go i think.

in fact it's not too hard, you can trigger "custom" videojs events in the source-handler here.

just call `trigger` function on the tech object. you don't need to declare the event anywhere in videojs before-hand. it can be an event that is only triggered by this source-handler. but trigger it via videojs, don't expose the hls.js stuff itself.

So basically in the source-handler here you'd say:

```
hls.on(Events.SOME_EVENT, function(event, data) { _tech.trigger('some-custom-hls-event', data); });
```

Something like that. see what i mean?

**Comment by tchakabam** [comment]
and then in your app:

```
player.tech().on('some-custom-hls-event', function() { ... })
```

**Comment by aquu** [comment]
Yeah, that sounds reasonable. I'm not so familiar on "how thing should be done" inside players. I just try to make things work outside the player :) Thanks for your @tchakabam !

**Comment by mkhazov** [comment]
I'd like to pass `autoStartLoad: false` in hlsjsConfig and then call `hls.startLoad()` when user starts the playback, to avoid loading the stream until it's needed.

I need to access Hls object for that, and I can accomplish this by  adding`tech._hls = _hls;` in `StreamrootProviderHLS.initialize()`.

Is there a better way to do this? What are the disadvantages of allowing to access the Hls object directly?

**Comment by tchakabam** [comment]
@mkhazov The stream should only start loading as you start playback actually. That is something that should be fixed in the recent versions. Did you try the most recent? If you see that the stream is loading before calling play the first time, you should probably file a bug (what browsers etc..).

However even if you would want to trigger loading "manually" such a behavior would rather be obtained by calling the `load` function on the videojs instance - ideally your application should only interact with videojs, and not with Hls.js as it is running wrapped in there.

**Comment by woutd** [comment]
@tchakabam When setting autoStartLoad in hlsjsConfig the stream does not load when starting playback. It is required to call hls.startLoad(), which is never done by videojs5-hlsjs-source-handler.

**Comment by tchakabam** [comment]
@woutd ah ok I understand the problem now, and what @mkhazov initially meant! sorry i was only reading over this quickly the other day. 

your PR seems to be a good fix for that :)

unfortunately i am not anymore a maintainer of this project with any write privileges. i just happen to being notified about these posts by chance. and i am always happy to answer any questions about anything when i can ;)

you should probably check with @ruslandinov if he can merge this into upstream master branch.

**Comment by ruslandinov** [comment]
Hey guys, sorry for late reply. 

We've updated the repo and added the `beforeinitialize` hook feature, that gives you access to hls.js instance right after it is created without actually exposing hls.js instance as a property. 

Please check this out https://github.com/streamroot/videojs-hlsjs-plugin#initialization-hook.  It should do the trick.
